lib/async_hooks.js

```diff
@@ -345,15 +346,10 @@ function emitInitS(asyncId, type, triggerAsyncId, resource) {
 
 function emitBeforeN(asyncId) {
   processing_hook = true;
-  // Use a single try/catch for all hook to avoid setting up one per iteration.
-  try {
-    for (var i = 0; i < active_hooks_array.length; i++) {
-      if (typeof active_hooks_array[i][before_symbol] === 'function') {
-        active_hooks_array[i][before_symbol](asyncId);
-      }
+  for (var i = 0; i < active_hooks_array.length; i++) {
+    if (typeof active_hooks_array[i][before_symbol] === 'function') {
+      active_hooks_array[i][before_symbol](asyncId);
```

Comment:
Allowing an exception to bubble then be handled by a Promise or similar means that it's possible that not all hook callbacks execute, but users that have those unexecuted callbacks won't know that it hasn't been called. Consider the following:
```js
const map = new Map();
createHook({
  before(id) { map.set(id, /* some object */) },
  after(id) { map.delete(id) },
}).enable();
```
You've just introduced a vector for a memory leak.